cdap-app-fabric/src/main/java/org/apache/twill/internal/yarn/AbstractYarnTwillService.java

```diff
@@ -0,0 +1,179 @@
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Map;
+
+/**
+ * TODO: copied from Twill 0.7 AbstractYarnTwillService for CDAP-5844.
+ * Remove after this fix is moved to Twill (TWILL-171).
+ *
+ * Abstract class for implementing {@link com.google.common.util.concurrent.Service} that runs in
+ * YARN container which provides methods to handle secure token updates.
+ */
+public abstract class AbstractYarnTwillService extends AbstractTwillService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractYarnTwillService.class);
+  protected final Location applicationLocation;
+  protected volatile Credentials credentials;
+
+  protected AbstractYarnTwillService(ZKClient zkClient, RunId runId, Location applicationLocation) {
+    super(zkClient, runId);
+    this.applicationLocation = applicationLocation;
+  }
+
+  /**
+   * Returns the location of the secure store, or {@code null} if either not running in secure mode or an error
+   * occur when trying to acquire the location.
+   */
+  protected final Location getSecureStoreLocation() {
+    if (!UserGroupInformation.isSecurityEnabled()) {
+      return null;
+    }
+    try {
+      return applicationLocation.append(Constants.Files.CREDENTIALS);
+    } catch (IOException e) {
+      LOG.error("Failed to create secure store location.", e);
+      return null;
+    }
+  }
+
+  /**
+   * Attempts to handle secure store update.
+   *
+   * @param message The message received
+   * @return {@code true} if the message requests for secure store update, {@code false} otherwise.
+   */
+  protected final boolean handleSecureStoreUpdate(Message message) {
+    if (!SystemMessages.SECURE_STORE_UPDATED.equals(message)) {
+      return false;
+    }
+
+    // If not in secure mode, simply ignore the message.
+    if (!UserGroupInformation.isSecurityEnabled()) {
+      return true;
+    }
+
+    try {
+      Credentials credentials = new Credentials();
+      Location location = getSecureStoreLocation();
+      try (DataInputStream input = new DataInputStream(new BufferedInputStream(location.getInputStream()))) {
+        credentials.readTokenStorageStream(input);
+      }
+
+      UserGroupInformation.getCurrentUser().addCredentials(credentials);
+
+      // CDAP-5844 Workaround for HDFS-9276, to update HDFS delegation token for long running application in HA mode
+      cloneHaNnCredentials(location, UserGroupInformation.getCurrentUser());
+      this.credentials = credentials;
+
+      LOG.info("Secure store updated from {}.", location);
+
+    } catch (Throwable t) {
+      LOG.error("Failed to update secure store.", t);
+    }
+
+    return true;
+  }
+
+  private static void cloneHaNnCredentials(Location location, UserGroupInformation ugi) throws IOException {
+    Configuration hConf = getConfiguration(location.getLocationFactory());
+    String scheme = getFileSystem(location.getLocationFactory(), hConf).getScheme();
```

Comment:
Hmm, this means that local location can't be passed to AbstractYarnTwillService, then?